component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthHandler.java

```diff
@@ -0,0 +1,190 @@
+    private String validIssuer;
+
+    private JWTValidator jwtValidator;
+
+    @Override
+    public void init(Properties properties) throws IdentityOAuth2Exception {
+        authConfig = properties.getProperty(OAuthConstants.CLIENT_AUTH_CREDENTIAL_VALIDATION);
+        initialiseJWTValidatorVariables(properties);
+        jwtValidator = new JWTValidator(notAcceptBeforeTimeInMins, preventTokenReuse, enableJTICache,
+                validAudience, validIssuer);
+    }
+
+    /**
+     * Initialize validator parameters
+     * @param properties
+     */
+    private void initialiseJWTValidatorVariables(Properties properties) {
+        try {
+            String rejectBeforePeriodConfigVal = properties.getProperty(Constants.VALIDITY_PERIOD);
+            if (isNotEmpty(rejectBeforePeriodConfigVal)) {
+                notAcceptBeforeTimeInMins = Integer.parseInt(rejectBeforePeriodConfigVal);
+            } else {
+                notAcceptBeforeTimeInMins = Constants.DEFAULT_VALIDITY_PERIOD_IN_MINUTES;
+            }
+            enableJTICache = Constants.DEFAULT_ENABLE_JTI_CACHE;
+            validAudience = Constants.DEFAULT_AUDIENCE;
+            validIssuer = Constants.DEFAULT_ISSUER;
+            preventTokenReuse = Constants.PREVENT_TOKEN_REUSE;
+
+        } catch (NumberFormatException e) {
+            log.warn("Invalid PrivateKeyJWT Validity period found in the configuration. Using default value.");
+            notAcceptBeforeTimeInMins = Constants.DEFAULT_VALIDITY_PERIOD_IN_MINUTES;
+        }
+    }
+
+    @Override
+    public boolean canAuthenticate(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+
+        String clientAssertionType = getRequestParameter(tokReqMsgCtx, Constants.OAUTH_JWT_ASSERTION_TYPE);
+        String clientAssertion = getRequestParameter(tokReqMsgCtx, Constants.OAUTH_JWT_ASSERTION);
+
+        return isValidJWTClientAssertionRequest(clientAssertionType, clientAssertion);
+        //TODO check validate SAML_Bearer
+    }
+
+    /**
+     * Check whether the client_assertion_type and client_assertion is valid
+     * @param clientAssertionType
+     * @param clientAssertion
+     * @return
+     */
+    private boolean isValidJWTClientAssertionRequest(String clientAssertionType, String clientAssertion) {
+        if (log.isDebugEnabled()) {
+            log.debug("Authenticate Requested with : " + Constants.OAUTH_JWT_ASSERTION_TYPE + " and " +
+                    "private_key_jwt: " + clientAssertion);
+        }
+        return Constants.OAUTH_JWT_BEARER_GRANT_TYPE.equals(clientAssertionType) && isNotEmpty(clientAssertion);
+    }
+
+    @Override
+    public boolean authenticateClient(OAuthTokenReqMessageContext tokReqMsgCtx)
+            throws IdentityOAuth2Exception {
+        try {
+            SignedJWT signedJWT = getSignedJWT(tokReqMsgCtx);
+            return jwtValidator.isValidToken(signedJWT);
+        } catch (IdentityOAuth2Exception e) {
+            return false;
```

Comment:
We need to handle this IdentityOAuth2Exception, if it's intentional please explain  through a comment

Reply:
Added an inline comment